### ISSUE TITLE
Public Yolov3 model get wrong predictions.

### ISSUE THREAD

**Issue by Ludwig-Zh** [openvinotoolkit/openvino#1593]
Hi,

   These days, follow the guild [https://docs.openvinotoolkit.org/latest/omz_models_public_yolo_v3_tf_yolo_v3_tf.html](url), I download a pre-trained yolov3 weights and convert it tensorflow .pb file ,then convert to IR by:
   "python3 mo_tf.py
--input_model /path/to/yolo_v3.pb
--tensorflow_use_custom_operations_config /path/to/yolo_v3.json
--batch 1"
  The json file is:
  "[
  {
    "id": "TFYOLOV3",
    "match_kind": "general",
    "custom_attributes": {
      "classes": 80,
      "anchors": [10, 13, 16, 30, 33, 23, 30, 61, 62, 45, 59, 119, 116, 90, 156, 198, 373, 326],
      "coords": 4,
      "num": 9,
      "masks":[[6, 7, 8], [3, 4, 5], [0, 1, 2]],
      "entry_points": ["conv2d_58/BiasAdd", "conv2d_66/BiasAdd", "conv2d_74/BiasAdd"]
    }
  }
]"
  After converting successfully, run ./object_detection_demo_yolov3_async -i <path_to_video>/inputVideo.mp4 -m <path_to_model>/yolo_v3.xml. Then got unexpected predictions like this:
  
![Detection results_001](https://user-images.githubusercontent.com/37934398/89140670-d84bf280-d574-11ea-8940-dde70004db93.png)

 On the contrary, use this  repository [https://github.com/mystic123/tensorflow-yolo-v3](url) which is the  demo for model converting for YOLOv3 , it works well for prediction. But  this repository do not has a train pipeline for user data training.

  So,  Is there a proved YOLOv3 implement for training of user-data, and it works well with OPENVINO? Or some tips on the unexpected predictions of the public model above?

  A lot of questions, thanks for your patience. 

  My environments:
  Ubuntu1804
  tensorflow1.14
  openvino2020.3

**Comment by jgespino** [comment]
Hi @Ludwig-Zh 

Thanks for reaching out! Please use the YOLOv3 implementation from pjreddie in the DarkNet framework:
YOLOv3 Weights: https://pjreddie.com/media/files/yolov3.weights
Coco names file: https://raw.githubusercontent.com/pjreddie/darknet/master/data/coco.names

Try following the instructions on this document: https://docs.openvinotoolkit.org/latest/openvino_docs_MO_DG_prepare_model_convert_model_tf_specific_Convert_YOLO_From_Tensorflow.html

The mystic123/tensorflow-yolo-v3 GitHub you linked should be used to convert the weight files to frozen TensorFlow model. If you want to train YOLOv3 on your own custom dataset, I recommend training in the DarkNet Framework. Take a look at the Training YOLO section in the following website, additionally there are guides online that walk you though the process.
https://pjreddie.com/darknet/yolo/

Hope this helps!

Regards,
Jesus

**Comment by Ludwig-Zh** [comment]
Thanks @jgespino

  I am sure the weights file was download from [https://pjreddie.com/media/files/yolov3.weights](url), and it worked well in the original keras project [https://github.com/david8862/keras-YOLOv3-model-set](url),coco names too.
 
  I checked up and found the similar issue was proposed in [https://community.intel.com/t5/Intel-Distribution-of-OpenVINO/Yolov3-from-public-models-gives-wrong-bounding-boxes/m-p/1140937](url). So maybe somewhere of the public implement of Yolov3  was incompatible with the Openvino toolkit.

  Train user data in darknet and convert to tensorflow by offical demo maybe a good ideal, I'll try later. Cause Yolov3 is  popular object detection solution these days, a smooth implement is helpful.
  
 Sincerely,
  Ludwig-Zh

**Comment by jgespino** [comment]
@Ludwig-Zh 

I tested the Yolo V3 model with our object_detection_demo_yolov3_async and saw accurate detections. In your **yolo_v3.json** you are using `"entry_points": ["conv2d_58/BiasAdd", "conv2d_66/BiasAdd", "conv2d_74/BiasAdd"]`. Try using the included configuration file under `/opt/intel/openvino/deployment_tools/extensions/front/tf/yolo_v3.json`. The entry points are different than your configuration file.

`"entry_points": ["detector/yolo-v3/Reshape", "detector/yolo-v3/Reshape_4", "detector/yolo-v3/Reshape_8"]`

Regards,
Jesus

**Comment by Ludwig-Zh** [comment]
@jgespino 

Thanks for reply. I tried but the entry points in public model is different  from official demo.So got an error msg:
"
[ ERROR ]  Exception occurred during running replacer "TFYOLOV3" (<class 'extensions.front.YOLO.YoloV3RegionAddon'>): TensorFlow YOLO V3 conversion mechanism was enabled. Entry points "detector/yolo-v3/Reshape, detector/yolo-v3/Reshape_4, detector/yolo-v3/Reshape_8" were provided in the configuration file. Entry points are nodes that feed YOLO Region layers. Node with name detector/yolo-v3/Reshape doesn't exist in the graph. Refer to documentation about converting YOLO models for more information."

Further, I inspect the pb file by Netron,maybe entry_points  ["conv2d_58/BiasAdd", "conv2d_66/BiasAdd", "conv2d_74/BiasAdd"] are correct prediction entries.This json file also download from public model zoo together with the pb file.

Sincerely,
Ludwig-Zh

**Comment by jgespino** [comment]
Hi @Ludwig-Zh
 
Thanks for the additional information, the YOLOv3 model from Open Model Zoo is different than the YOLOv3 used in the documentation. Using the model from the open model zoo, the entry points are `"entry_points": ["conv2d_58/BiasAdd", "conv2d_66/BiasAdd", "conv2d_74/BiasAdd"]`.

The boundary boxes having an incorrect size is likely due to the model optimizer command. Try using the two commands below to download and convert the model/ The download command will download the model and configuration file (.json).
```
python3 /opt/intel/openvino/deployment_tools/tools/model_downloader/downloader.py --name yolo-v3-tf
python3 /opt/intel/openvino/deployment_tools/tools/model_downloader/converter.py --name yolo-v3-tf
```

The converter script will convert the model with the following MO command:
```
========== Converting yolo-v3-tf to IR (FP16)
Conversion command: /usr/bin/python3 -- /opt/intel/openvino_2020.4.287/deployment_tools/model_optimizer/mo.py --framework=tf --data_type=FP16 --output_dir=/home/jesus/cases/1593-public-yolov3-wrng-predictions/omz/public/yolo-v3-tf/FP16 --model_name=yolo-v3-tf '--input_shape=[1,416,416,3]' --input=input_1 '--scale_values=input_1[255]' --reverse_input_channels --transformations_config=/home/jesus/1593/omz/public/yolo-v3-tf/yolo-v3.json --input_model=/home/jesus/1593/omz/public/yolo-v3-tf/yolo-v3.pb

========== Converting yolo-v3-tf to IR (FP32)
Conversion command: /usr/bin/python3 -- /opt/intel/openvino_2020.4.287/deployment_tools/model_optimizer/mo.py --framework=tf --data_type=FP32 --output_dir=/home/jesus/1593/omz/public/yolo-v3-tf/FP32 --model_name=yolo-v3-tf '--input_shape=[1,416,416,3]' --input=input_1 '--scale_values=input_1[255]' --reverse_input_channels --transformations_config=/home/jesus/1593/omz/public/yolo-v3-tf/yolo-v3.json --input_model=/home/jesus/cases/1593-public-yolov3-wrng-predictions/omz/public/yolo-v3-tf/yolo-v3.pb
```

Give that a try and let me know if it works for you.

Regards,
Jesus

**Comment by Ludwig-Zh** [comment]
@jgespino 

  Thanks! I tried the MO command, and it works! Further more, I use this script to convert the public model generated by keras [https://docs.openvinotoolkit.org/latest/omz_models_public_yolo_v3_tf_yolo_v3_tf.html](url), it works too by simple change of "--input image_input".

 So the key issue is the MO command parameters.

**Comment by jgespino** [comment]
@Ludwig-Zh Great, glad that worked for you! Please let me know if we can proceed to close this issue.

**Comment by Ludwig-Zh** [comment]
@jgespino ok, this issue can be closed.
